lib/src/main/java/io/stormbird/token/tools/TokenScriptValidator.java

```diff
@@ -0,0 +1,261 @@
+        NodeList nl = xml.getElementsByTagNameNS(XMLSignature.XMLNS, "Signature");
+        if (nl.getLength() == 0) return result;
+
+        // Create a DOM XMLSignatureFactory that will be used to unmarshal the
+        // document containing the XMLSignature
+        XMLSignatureFactory fac = XMLSignatureFactory.getInstance("DOM");
+
+        // Create a DOMValidateContext and specify a KeyValue KeySelector
+        // and document context
+        DOMValidateContext valContext = new DOMValidateContext(new SigningCertSelector(), nl.item(0));
+
+        // unmarshal the XMLSignature
+        XMLSignature signature = fac.unmarshalXMLSignature(valContext);
+
+        // Validate the XMLSignature (generated above)
+        result.isValid = signature.validate(valContext);
+
+        //check that the tsml file is signed by a valid certificate
+        return validateSSLCertificateIssuer(signature, result);
+    }
+
+    private static TSMLValidationResult validateSSLCertificateIssuer(XMLSignature signature, TSMLValidationResult result)
+    {
+        try
+        {
+            KeyInfo xmlKeyInfo = signature.getKeyInfo();
+            List certList = getCertificateChainFromXML(xmlKeyInfo.getContent());
+            X509Certificate signingCert = selectSigningKeyFromXML(xmlKeyInfo.getContent());
+            boolean validIssuer = validateSSLCertificateBasedOnTrustedCA(certList);
+            if (result.isValid && validIssuer)
+            {
+                result.issuerPrincipal = signingCert.getIssuerX500Principal().getName();
+                result.subjectPrincipal = signingCert.getSubjectX500Principal().getName();
+                result.keyType = signingCert.getSigAlgName();
+                for (Object o : xmlKeyInfo.getContent())
+                {
+                    XMLStructure xmlStructure = (XMLStructure) o;
+                    if (xmlStructure instanceof KeyName)
+                    {
+                        result.keyName = ((KeyName) xmlStructure).getName();
+                    }
+                }
+            }
+            else
+            {
+                result.isValid = false;
+            }
+        }
+        catch(Exception e)
+        {
+            result.isValid = false;
+            result.failureReason = e.getMessage();
+        }
+        return result;
+    }
+
+    private static boolean validateSSLCertificateBasedOnTrustedCA(List<X509Certificate> certList) throws Exception
+    {
+        // By default on Oracle JRE, algorithm is PKIX
+        TrustManagerFactory tmf = TrustManagerFactory
+                .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        // 'null' will initialise the tmf with the default CA certs installed
+        // with the JRE.
+        tmf.init((KeyStore) null);
+        X509TrustManager tm = (X509TrustManager) tmf.getTrustManagers()[0];
+        X509Certificate[] validIssuers = tm.getAcceptedIssuers();
+        int index = 0;
+        while(certList.size() > index)
```

Comment:
Yes we should, @James-Sangalli There is not much performance punishment if you use an algorithm at O² efficiency so you can come up with something stupid.

Reply:
Hi, I believe you should be able to validate the cert chain using `CertPathValidator`: 

https://docs.oracle.com/javase/8/docs/api/java/security/cert/CertPathValidator.html#validate-java.security.cert.CertPath-java.security.cert.CertPathParameters-

Here is a code snippet of an alternative to the current `validateSSLCertificateBasedOnTrustedCA` method, that demonstrates the use of `CertPathValidator`:

https://gist.github.com/n1v0lg/4847b7821876412ebc5cf967070f8be3

I haven't been able to validate it works because the code currently fails with a Duplicate KeyValue error before it gets to this portion of the code.

Reply:
> this approach requires that the certificate chain is correctly ordered, i.e., the certs are in reverse order of issuance (root CA comes last).

Is there a convention of such? If so, we can simply require them to be ordered in the documents of creation of such signatures.

Reply:
Seems to attest here that it is supposed to be order like that but in practise many clients can verify it unordered. 

https://utcc.utoronto.ca/~cks/space/blog/tech/SSLChainOrder

Reply:
I think it would be ideal to order the certs as part of the validation, instead on relying that they are ordered. 

Basically, you just need to find the root cert, then the cert it signed, and so on by matching up `getSubjectDN` with `getIssuerDN` (see here for example implementation: https://www.programcreek.com/java-api-examples/?code=pinguo-sunjianfei/Android-Application-ZJB/Android-Application-ZJB-master/extras/zjb-sync-volley/src/main/java/com/zjb/volley/core/ssl/GenericX509TrustManager.java).

Reply:
@n1v0lg added the code, seems to be working and pretty straightforward with matching by subject and issuer.